Support ADC driver for Renesas RA8, RA6, RA4 devices

khoa-nguyen-18 opened the pull request:
- Add missing ADC document for ek_ra8m1 board
- Remove test id for ek_ra8m1 in the `adc_accuracy_test`
- Add 2 ADC properties in Renesas RA ADC node
- Enable ADC support on these boards:
RA8: ek_ra8d1, mck_ra8t1
RA6: ek_ra6m1, ek_ra6m2, ek_ra6m3, ek_ra6m4, ek_ra6m5, ek_ra6e2, fpb_ra6e1, fpb_ra6e2,
RA4: ek_ra4m2, ek_ra4m3, ek_ra4e2, ek_ra4w1

khoa-nguyen-18 committed:
drivers: adc: Add ADC properties in Renesas RA ADC node

- Add "channel-available-mask" property in ADC node
to detect which channels are available to use

- Add "add-average-count" property in ADC node to chose
number of count of the addition or average mode

- Change the source code of ADC to match with 2 new properties.

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 commented:
Add a commit to: 
- Support ADC RA8D1, RA8T1
- Add the missing ADC doc for EK-RA8M1
- Remove the test id device of EK-RA8M1 in testcase.yaml of adc_accuracy_test

khoa-nguyen-18 commented:
Rebase main and solve the conflicts

soburi commented on drivers/adc/adc_renesas_ra.c:
```diff
@@ -313,7 +314,7 @@ static int adc_ra_init(const struct device *dev)
 
 const adc_extended_cfg_t g_adc_cfg_extend = {
-	.add_average_count = ADC_ADD_OFF,
+	.add_average_count = ADC_ADD_AVERAGE_FOUR,
```
This property seems should be better to make configurable with devicetree.

soburi commented on boards/renesas/ek_ra8d1/doc/index.rst:
```diff
@@ -110,6 +110,8 @@ The below features are currently supported on Zephyr OS for EK-RA8D1 board:
 +--------------+------------+------------------+
 | CAN          | on-chip    | canfd            |
 +--------------+------------+------------------+
+| ADC          | on-chip    | adc              |
```
Wrong commmit message. This shoud start with `boards: renesas: ...`

khoa-nguyen-18 replied:
Thank you, I have changed the commit message and split this modification into a commit

soburi commented on boards/renesas/ek_ra6e2/doc/index.rst:
```diff
@@ -98,6 +98,8 @@ The below features are currently supported on Zephyr OS for EK-RA6E2 board:
 +-----------+------------+----------------------+
 | CLOCK     | on-chip    | clock control        |
 +-----------+------------+----------------------+
+| ADC       | on-chip    | adc                  |
```
Wrong commmit message. This shoud start with `boards: renesas: ...`

khoa-nguyen-18 replied:
also with this issue

soburi commented on dts/arm/renesas/ra/ra6/r7fa6e10x.dtsi:
```diff
@@ -7,6 +7,8 @@
 #include <arm/renesas/ra/ra6/ra6-cm33-common.dtsi>
 #include <zephyr/dt-bindings/clock/ra_clock.h>
 
+/delete-node/ &adc1;
```
It's better to split the changes in the dts folder into a separate commit.

khoa-nguyen-18 committed:
boards: renesas: Add missing document for ADC in EK-RA8M1

- Add missing description for ADC in EK-RA8M1 document

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 committed:
dts: arm: renesas: Add dts node to support ADC for RA6, RA4

Add dts node to support canfd for RA6, RA4 MCU: r7fa6m5xh,
r7fa6m4ax, r7fa6m3ax, r7fa6m2ax, r7fa6m1ad3cfp, r7fa6e10x,
r7fa6e2bx, r7fa4w1ad2cng, r7fa4m3ax, r7fa4m2ax, r7fa4e2b93cfm

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 committed:
boards: renesas: Add ADC support for Renesas RA8, RA6, RA4

Add ADC support for EK-RA8D1, MCK-RA8T1, EK-RA6M1, EK-RA6M2,
EK-RA6M3, EK-RA6M4, EK-RA6M5, EK-RA6E2, FPB-RA6E1, FPB-RA6E2,
EK-RA4M2, EK-RA4M3, EK-RA4E2, EK-RA4W1

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 committed:
tests: drivers: adc: Support RA8, RA6, RA4 to test accuracy and api

Add overlay to support test adc_accuracy_test and adc_api in boards:
EK-RA8D1, MCK-RA8T1, EK-RA6M1, EK-RA6M2, EK-RA6M3, EK-RA6M4, EK-RA6M5,
EK-RA6E2, FPB-RA6E1, FPB-RA6E2, EK-RA4M2, EK-RA4M3, EK-RA4E2, EK-RA4W1

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 committed:
tests: drivers: adc: Remove test id for EK-RA8M1

- Remove test id device of ek_ra8m1 in testcase.yaml of
adc_accuracy_test

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 commented:
Remove redundant change

soburi reviewed (dismissed):
LGTM

khoa-nguyen-18 commented:
I just rebased main to solve conflicts, @soburi @anangl can you please take a look?

khoa-nguyen-18 commented:
Just rebase main to solve the conflicts

khoa-nguyen-18 commented:
just rebase main to solve twister test fail

thaoluonguw reviewed (dismissed):
It looks good to me.

soburi commented on boards/renesas/ek_ra4e2/ek_ra4e2.dts:
```diff
@@ -82,3 +83,9 @@
 	pinctrl-names = "default";
 	status = "okay";
 };
+
```
Please Fix CI isseus.

khoa-nguyen-18 replied:
I have fixed the CI issues, thank you

khoa-nguyen-18 commented:
@anangl Can you please take a look?

khoa-nguyen-18 commented on tests/drivers/adc/adc_accuracy_test/boards/ek_ra4e2.overlay:
As per current policy, we specifically update the overlay for each board. This helps us make fewer decisions or take care of future common/private boards.

khoa-nguyen-18 commented on tests/drivers/adc/adc_api/boards/ek_ra6m1.overlay:
same as about, thank you for your suggestion

khoa-nguyen-18 commented on dts/bindings/adc/renesas,ra-adc.yaml:
```diff
@@ -24,5 +24,18 @@ properties:
+    required: true
+    description: Mask for ADC channels existed in each board
+
+  average-count:
+    type: int
+    default: 1
+    enum: [1, 2, 4, 8, 16]
+    description: The conversion count of the average mode
+      The ADC module will take multiple samples of an analog signal
+      and averages them to reduce noise and enhance accuracy
```
It can't because this property which we use for the ADC **unit** configuration, and the `oversampling` is for [**child** node](https://github.com/zephyrproject-rtos/zephyr/blob/3f60489fae6445439d9eecb67b95af3636e78ab8/dts/bindings/adc/adc-controller.yaml#L20)

bjarki-andreasen commented:
ping @soburi
